core/src/main/java/com/google/instrumentation/trace/SpanId.java

```diff
@@ -13,33 +13,80 @@
 
 package com.google.instrumentation.trace;
 
+import static com.google.common.base.Preconditions.checkArgument;
+
 import com.google.common.base.MoreObjects;
-import com.google.common.base.Objects;
+import com.google.common.io.BaseEncoding;
 
-import java.util.Formatter;
-import java.util.Locale;
+import java.util.Arrays;
+import java.util.Random;
 import javax.annotation.concurrent.Immutable;
 
 /**
- * A class that represents a span identifier. A span identifier is a 64-bit, unsigned integer.
+ * A class that represents a span identifier. A valid span identifier is a 8-bytes array with
+ * at least one non-zero byte.
  */
 @Immutable
-public final class SpanId {
-  private static final long INVALID_SPAN_ID = 0;
-  private final long spanId;
+public final class SpanId implements Comparable<SpanId> {
+  // The size in bytes of the span id.
+  private static final int SPAN_ID_SIZE = 8;
+  private final byte[] bytes;
+
+  private SpanId(byte[] bytes) {
+    this.bytes = bytes;
+  }
+
+  /**
+   * The invalid {@code SpanId}. All bytes are 0.
+   */
+  public static final SpanId INVALID = new SpanId(new byte[SPAN_ID_SIZE]);
+
+  /**
+   * Returns a {@code SpanId} whose representation is given param.
+   *
+   * @param bytes the representation of the {@code SpanId}.
+   * @return a {@code SpanId} whose representation is given param.
+   * @throws NullPointerException if bytes is null.
+   * @throws IllegalArgumentException if bytes length is not 8.
+   */
+  public static SpanId fromBytes(byte[] bytes) {
+    checkArgument(bytes.length == SPAN_ID_SIZE, "bytes");
+    byte[] bytesCopy = Arrays.copyOf(bytes, SPAN_ID_SIZE);
+    return Arrays.equals(bytesCopy, INVALID.bytes) ? INVALID : new SpanId(bytes);
+  }
+
+  /**
+   * Generates a new random {@code SpanId}.
+   *
+   * @param random The random number generator.
+   * @return a valid new {@code SpanId}.
+   */
+  public static SpanId generateRandomId(Random random) {
```

Comment:
Was Random introduced in Java SE 7? If so, what do we do to support earlier versions?